Sources/Commands/SwiftScriptTool.swift

```diff
@@ -0,0 +1,374 @@
+    public func run() throws {
+        guard let file = options.file else {
+            throw ScriptError.fileNotFound("")
+        }
+        let (productName, cacheDirPath) = try checkAndPerformCache(for: file)
+
+        var swiftOptions = swiftOptions
+        swiftOptions.packagePath = cacheDirPath
+        swiftOptions.buildPath = nil
+        let swiftTool = try SwiftTool(options: swiftOptions)
+
+        try self.run(swiftTool, as: productName, at: cacheDirPath)
+        if swiftTool.diagnostics.hasErrors || swiftTool.executionStatus == .failure {
+            throw ExitCode.failure
+        }
+    }
+
+    public static var _errorLabel: String { "error" }
+}
+
+struct ScriptToolOptions: ParsableArguments {
+    /// If the executable product should be built before running.
+    @Flag(name: .customLong("skip-build"), help: "Skip building the executable product")
+    var shouldSkipBuild: Bool = false
+
+    var shouldBuild: Bool { !shouldSkipBuild }
+
+    /// The script file to run.
+    @Argument(help: "The script file to run")
+    var file: String?
+
+    /// The arguments to pass to the executable.
+    @Argument(parsing: .unconditionalRemaining,
+              help: "The arguments to pass to the executable")
+    var arguments: [String] = []
+}
+
+/// swift-script tool namespace
+public struct SwiftScriptTool: ParsableCommand {
+    public static var configuration = CommandConfiguration(
+        commandName: "script",
+        _superCommandName: "swift",
+        abstract: "Manage and run Swift scripts",
+        discussion: "SEE ALSO: swift build, swift run, swift package, swift test",
+        version: SwiftVersion.currentVersion.completeDisplayString,
+        subcommands: [
+            Run.self,
+            Build.self,
+            Clean.self,
+            Reset.self,
+            Resolve.self,
+            List.self,
+        ],
+        helpNames: [.short, .long, .customLong("help", withSingleDash: true)])
+
+    public init() {}
+
+    public static var _errorLabel: String { "error" }
+}
+
+/// swift-run tool namespace
+extension SwiftScriptTool {
+    struct Run: ScriptCommand {
+        static let configuration = CommandConfiguration(
+            abstract: "Runs a script")
+
+        @OptionGroup(_hiddenFromHelp: true)
+        var swiftOptions: SwiftToolOptions
+
+        @OptionGroup()
+        var options: ScriptToolOptions
+
+        /// Whether to print build progress.
+        @Flag(help: "Print build progress")
+        var quiet: Bool = false
+
+        func run(_ swiftTool: SwiftTool, as productName: String, at cacheDirPath: AbsolutePath) throws {
+            let output = BufferedOutputByteStream()
+            if quiet {
+                swiftTool.redirectStdoutTo(.init(output))
+            } else {
+                swiftTool.redirectStdoutToStderr()
+            }
+            // FIXME: More elegant solution?
+            print(diagnostic: .init(message: .note("Using cache: \(cacheDirPath.basename)")),
```

Comment:
If `quiet` is set, this will have already been redirected to a buffer that won't be displayed — am I understanding that correctly?  In other words, it won't be shown when quiet?

I think that quiet should perhaps be the default, or to turn it around, that there should be a `verbose` flag that can provide extra information (such as whether cache is used).

Reply:
Also: is it the case that this only says whether a cache _could_ be used?  It doesn't seem to indicate whether a cache result was actually found, or am I reading that incorrectly?

Reply:
> If quiet is set, this will have already been redirected to a buffer that won't be displayed

Yes. Except that the build fails.

> Also: is it the case that this only says whether a cache could be used? It doesn't seem to indicate whether a cache result was actually found, or am I reading that incorrectly?

The cache name displayed here acts as a shortcut to the file.eg., you can call `swift script run test-67a8a2` instead of `swift script run /Users/yr.chen/Developer/GSoC/test.swift` once the cache is set up.

Currently it cannot detect whether there was an up-to-date prebuild. We can add this feature and emit building step (and its outputs) when a cache is exactly hit, making `clean` as a default if the script was prebuilt.